### PR TITLE
Perform global deinitialization when exiting the game launcher

### PR DESCRIPTION
**Pull request by burelc-amzn**
This removes the `TerminateProcess()` / `exit()` calls from `CSystem::Quit`, and fixes all the crashes and leaks that were occurring as a result of the atexit code now running.

### CODE REVIEW

File: Code/LauncherUnified/Launcher.cpp [hunk 1]
```diff
@@ -232,13 +233,17 @@ namespace
...
+        while (!gameApplication.WasExitMainLoopRequested())
         {
             // Pump the system event loop
             gameApplication.PumpSystemEventLoopUntilEmpty();
 
+            if (gameApplication.WasExitMainLoopRequested())
+            {
+                break;
+            }
+
```
Comment (lumberyard-employee-dm) [inline Code/LauncherUnified/Launcher.cpp:246]: This seems kind of "dowhily" for me. i.e `do ... while`

What if we pumped the System Event loop at the end this block instead?

File: Code/LauncherUnified/Launcher.cpp [hunk 2]
```diff
@@ -697,6 +699,11 @@ namespace O3DELauncher
             status = ReturnCode::ErrCrySystemInterface;
         }
 
+        if ((gEnv) && (gEnv->pSystem) && (gEnv->pSystem->GetILevelSystem()))
+        {
+            gEnv->pSystem->GetILevelSystem()->UnloadLevel();
+        }
```
Comment (lumberyard-employee-dm) [inline Code/LauncherUnified/Launcher.cpp:705]: Why not have the LevelSystem destructor do this?

File: Code/Legacy/CrySystem/AZCoreLogSink.h [hunk 3]
```diff
@@ -31,6 +31,14 @@ class AZCoreLogSink
     : public AZ::Debug::TraceMessageBus::Handler
 {
 public:
+    ~AZCoreLogSink()
+    {
+        if (BusIsConnected())
+        {
+            Disconnect();
+        }
```
Comment (lumberyard-employee-dm) [inline Code/Legacy/CrySystem/AZCoreLogSink.h:39]: It is always safe to unconditionally `Disconnect()` from an EBus.
It is the other way that causes Asserts. Connecting without checking whether the bus was connected.

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Code/LauncherUnified/Launcher.cpp (hunk 1 above)
Comment (bosnichd) [reply Code/LauncherUnified/Launcher.cpp:246]: Pumping the system event loop should be the first thing we do every frame. We could maybe pump it once before entering the loop, then at the end as you suggest, and the behavior would remain the same, but conceptually I think it's important for anyone looking at it to understand that it's the first thing to happen every frame, so I'd prefer leaving it as is.

**Review by AMZN-ScottR — APPROVED** [review]

File: Code/LauncherUnified/Launcher.cpp (hunk 2 above)
Comment (bosnichd) [reply Code/LauncherUnified/Launcher.cpp:705]: +1 - ideally we don't want to be adding any additional dependencies on the level system from anywhere considering we're trying to get rid of it.

File: Code/Legacy/CrySystem/System.cpp [hunk 4]
```diff
@@ -548,27 +548,12 @@ void CSystem::Quit()
...
 #define AZ_RESTRICTED_SECTION SYSTEM_CPP_SECTION_4
 #include AZ_RESTRICTED_FILE(System_cpp)
 #endif
 #if defined(AZ_RESTRICTED_SECTION_IMPLEMENTED)
 #undef AZ_RESTRICTED_SECTION_IMPLEMENTED
-#elif defined(WIN32) || defined(WIN64)
-    TerminateProcess(GetCurrentProcess(), m_env.retCode);
-#else
-    exit(m_env.retCode);
 #endif
```
Comment (bosnichd) [inline Code/Legacy/CrySystem/System.cpp:557]: Can you please remove all this while you're here? It's safe to do so now that you've got it shutting down correctly.

**Review by bosnichd — APPROVED** [review]
Awesome, thanks!

File: Code/LauncherUnified/Launcher.cpp (hunk 1 above)
Comment (burelc-amzn) [reply Code/LauncherUnified/Launcher.cpp:246]: I'm not sure under what conditions the application can be requested to exit. If `Tick()` can also call `Application::ExitMainLoop()`, and that frees resources that `PumpSystemEventLoopUntilEmpty()` expects to be valid, I think we'd be right back in the same situation. So I was erring on the side of caution here.

File: Code/Legacy/CrySystem/System.cpp (hunk 4 above)
Comment (burelc-amzn) [reply Code/Legacy/CrySystem/System.cpp]: I removed the `AZ_RESTRICTED_PLATFORM` stuff from this area as well.

File: Code/Legacy/CrySystem/AZCoreLogSink.h (hunk 3 above)
Comment (burelc-amzn) [reply Code/Legacy/CrySystem/AZCoreLogSink.h]: Fixed.

File: Code/LauncherUnified/Launcher.cpp (hunk 2 above)
Comment (burelc-amzn) [reply Code/LauncherUnified/Launcher.cpp]: Fixed.

File: Code/LauncherUnified/Launcher.cpp [hunk 5]
```diff
@@ -26,6 +26,7 @@
 
 #include <CryLibrary.h>
 #include <ISystem.h>
+#include <ILevelSystem.h>
```
Comment (bosnichd) [inline Code/LauncherUnified/Launcher.cpp:29]: Nit-pick: can we remove the #include as well?

**Review by bosnichd — APPROVED** [review]
Thanks!

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Code/LauncherUnified/Launcher.cpp (hunk 5 above)
Comment (burelc-amzn) [reply Code/LauncherUnified/Launcher.cpp]: D'oh, fixed.

**Review by amznestebanpapp — APPROVED** [review]

**Review by spham-amzn — APPROVED** [review]

**Comment by hwwxj** [comment]
hi, I am using the branch of 2107 in linux, and when I try to merge this change to my local branch.  and then I try to exit the game launcher, here comes the crash. do u have any ideas about this?
image

**Comment by amznestebanpapp** [comment]
> hi, I am using the branch of 2107 in linux, and when I try to merge this change to my local branch. and then I try to exit the game launcher, here comes the crash. do u have any ideas about this? image

Looks like a bug, LyShine seems to not be de-initialized properly (maybe a missing disconnect from the tick bus)

**Comment by hwwxj** [comment]
yeah, I think maybe the reason of this crash is that the function "CSystem::Release" is not called, which will cause the event "OnCrySystemShutdown" is not broadcast.     the order is like this :  CSystem::Release()->CSystem::~CSystem()->CSystem::ShutDown()->OnCrySystemShutdown->LyShineSystemComponent::OnCrySystemShutdown()->CLyShine::~CLyShine()->BusDisconnect()

**Pull request merged**